### PR TITLE
Disable in-process parallel test execution

### LINKED ISSUE #54967: Disable in-process parallel test execution
# Proposal: Disable In-Process Parallel Test Execution

## Summary

This proposal recommends disabling in-process parallel test execution in `dotnet/sdk`. The current approach of running tests concurrently within a single process is causing persistent CI instability that imposes a significant tax on the team's productivity.

## Intermittent Test Failures Are Hindering PR Velocity

Flaky tests caused by in-process parallelism are the single biggest source of friction in the PR merge workflow today.

### Obscuring signal and creating toil

When a PR shows red CI, you cannot tell at a glance whether the failure is something you introduced or a pre-existing intermittent flake. To find out, you have to dig into every failure, examine the test, check if it's a known issue, and compare against other recent runs. This investigation takes time, creates frustration, and is a constant tax on every PR author and reviewer.

The `/ba-g` (build analysis gate bypass) mechanism exists specifically to work around this, but its widespread use is a symptom of the problem, not a solution.

### Masking real regressions

Because of the volume of noise, real regressions occasionally slip through unnoticed. When every CI run has failures, it becomes statistically inevitable that some genuine issues are overlooked. Not because people are negligent, but because the signal is drowned out by the noise.

### Wasting resources through retries

When tests fail due to flakiness rather than a real problem, contributors retry CI, sometimes many times, hoping for a green run. This burns compute resources that are already in short supply and delays merges.

### The data

Of the **100 most recently merged PRs** into `main` (2026-06-12 to 2026-06-23):

| Classification | Percentage |
|:---------------|:-----------|
| Truly clean (no failures at all) | 37% |
| Known build errors auto-handled | 18% |
| Used `/ba-g` to bypass failures | 36% |
| Green only after retries | 5% |
| Failed CI, no `/ba-g` | 1% |
| Other (neutral) | 3% |

**Key takeaways:**
- **Only 37%** of PRs merge with a genuinely clean CI run
- **61% encounter test failures** that are either bypassed, auto-dismissed, retried away, or force merged

> See [Appendix: PR Data](#appendix-pr-data) for the full list of 100 PRs with links, CI results, and classification.

## CI on `main` Is Consistently Red

The `dotnet-sdk-public-ci` pipeline on `main` has a **3.7% pass rate** over the last ~30 days (2026-05-25 to 2026-06-23): only 4 out of 108 builds succeeded. The last green build was on **June 1, 2026**.

A consistently red `main` branch means:
- **No reliable baseline** to compare PR results against
- **Regressions go undetected** because there's no "known good" state

## Known Build Errors Are Underutilized

There are currently [23 open Known Build Error issues](https://github.com/dotnet/sdk/issues?q=is%3Aissue+is%3Aopen+label%3A%22Known+Build+Error%22) and [88 total filed historically](https://github.com/dotnet/sdk/issues?q=is%3Aissue+label%3A%22Known+Build+Error%22). However, the 35% `/ba-g` rate above shows that the true scope of intermittent failures far exceeds what is documented.

The Known Build Error system is underutilized because:
- Defining a correct regex pattern requires careful effort and there is no tooling to assist
- Overly broad patterns risk matching unrelated failures, which is dangerous
- It is simply easier to `/ba-g` and move on

A recent focused effort to properly document Known Build Errors has resulted in **17 issues filed in June 2026 alone**, compared to 0-2 per month for most of the prior year. This uptick is the result of actively investigating and cataloging the problem, not a sudden increase in flakiness.

## Root Cause: Writing Concurrent Test Code Is Difficult

Writing test infrastructure that correctly supports concurrency is inherently difficult. The issues are endless and often difficult to diagnose. The test failures are not product code bugs. The problem is that there are no guardrails that promote writing concurrency-safe test code. It is easy for test authors to follow existing patterns that work in isolation but fail under parallel execution, and these issues are rarely surfaced locally. Retrofitting concurrency safety across the test codebase is an unbounded effort.

### Common patterns

| Pattern | Description |
|---------|-------------|
| Shared file contention | Multiple tests write to the same temp files, binlogs, or solution files |
| Resource cleanup races | One test cleans up shared resources while another is still using them |
| Thread pool starvation | Leaked threads from watchers/pollers exhaust the thread pool |
| Timing-dependent assertions | Tests rely on filesystem notifications or async coordination that breaks under load |
| Directory traversal pollution | Tests discover files created by concurrent tests via parent directory walks |

### Illustrative example: [#54942](https://github.com/dotnet/sdk/issues/54942)

`PostActionTests.AddProjectToSolution_WithOutputRelativePath` fails intermittently because:

1. Test A creates a `.sln` file at the sandbox root directory using a temporary name (`GetTempName()`)
2. Test B, working in a subdirectory, discovers that `.sln` via solution discovery walking up the directory tree
3. Test A finishes and deletes its `.sln`
4. Test B tries to operate on the now-deleted `.sln` . **Failure**

Key observations:
- **The failing test is doing nothing wrong.** It's a different concurrent test that creates the conflict.
- **Diagnosis is extremely difficult.** Resources use `GetTempName()` rather than being named in context of the owning test. There is no way to trace an artifact back to the test that created it. You cannot identify the offending test without first refactoring how artifacts are named across the codebase.
- **These are not rare edge cases.** This pattern repeats across the entire test suite with variations on the theme.

### Additional examples

| Issue | Problem |
|-------|---------|
| [#54942](https://github.com/dotnet/sdk/issues/54942) | Solution file at sandbox root found by concurrent test via directory traversal, then deleted |
| [#54817](https://github.com/dotnet/sdk/issues/54817) | `IOException` from concurrent access to `msbuild.binlog` file |
| [#54176](https://github.com/dotnet/sdk/issues/54176) | Leaked `PollingDirectoryWatcher` threads cause thread pool starvation and hangs |
| [#53661](https://github.com/dotnet/sdk/issues/53661) | Hot-reload file change detection races assertion timeout under concurrent load |
| [#53870](https://github.com/dotnet/sdk/issues/53870) | Locked-file warning emission is timing-sensitive; sometimes sees 0 warnings |
| [#44878](https://github.com/dotnet/sdk/issues/44878) | VS instance discovery has unintended concurrent access to a critical section |
| [#9673](https://github.com/dotnet/sdk/issues/9673) | Restore tests had to serialize because parallel network operations caused flakiness |

### Why these are not fixable at scale

- **Shared global/process resources:** Files, caches, watchers, thread pool, VS discovery, all shared by default.
- **Async timing depends on machine load:** Filesystem events, thread scheduling, and network operations behave differently under contention. These failures rarely reproduce locally, which is how the problem grew undetected for so long.
- **Tests assume isolation:** Many tests were written assuming sequential execution; they share implicit state without knowing it.
- **Diagnosis requires refactoring:** You cannot identify the offending test without first making all test artifacts traceable, which is itself a large-scale refactoring effort.
- **No return on investment:** Time spent making test infrastructure concurrency-safe is time not spent improving the product or adding meaningful test coverage.

## Precedent: Roslyn's Approach

The Roslyn team (`dotnet/roslyn`) faced the exact same problem and made the deliberate decision to **disable in-process parallelism** years ago. Their approach is well-established and battle-tested.

### What Roslyn does

1. **Disabled xUnit parallelization**: their `eng/config/xunit.runner.json` sets `"parallelizeTestCollections": false`
2. **Multi-process local execution**: their `RunTests.dll` tool launches multiple subprocesses (`1.5 x processor count`), each running a test assembly in isolation
3. **Helix distribution for CI**: test assemblies are partitioned by historical execution time and distributed across multiple Helix machines

### Why Roslyn disabled parallelism

From [roslyn#80095](https://github.com/dotnet/roslyn/issues/80095):
> "Naive single-process parallelization wasn't tractable; workspace APIs and native components were problematic."

From [roslyn#321](https://github.com/dotnet/roslyn/issues/321):
> "xUnit 2.0 will run tests in parallel and several of our tests fail in that mode. Disabling for now."

The pattern is identical to what `dotnet/sdk` is experiencing today.

### Roslyn's results

By separating parallelism from isolation (run many processes, each sequential), Roslyn achieves:
- **Fast CI** through Helix distribution across machines
- **Fast local runs** through multi-process execution
- **Reliable tests** because each process has full isolation
- **No whack-a-mole**: no need to chase concurrency bugs in test infrastructure

### Immediate action: Disable in-process parallelism

1. Set `parallelizeTestCollections: false` in xUnit runner configuration
2. Set `MSTestParallelizeScope: None` for MSTest projects
3. Validate CI stability improves

This disables running multiple tests concurrently within a single process. Tests that are themselves written to exercise concurrency in the product (e.g., testing parallel workloads) are unaffected, as their internal concurrency is not controlled by the test runner.

### Expected outcomes

- **Dramatic reduction in intermittent test failures** (expected 90%+) as cross-test contamination is eliminated by isolation. A small number of tests with internal concurrency issues (e.g., [#54960](https://github.com/dotnet/sdk/issues/54960)) will remain, but the vast majority of current flakiness is caused by inter-test parallelism.
- **CI pass rate recovers** from 3.7% toward a healthy baseline
- **`/ba-g` usage drops** as flaky tests stabilize
- **Known Build Error backlog shrinks** as root causes are eliminated by isolation
- **Developer productivity improves**: less time retrying CI, investigating flakes, filing issues

### Cost

The primary concern is **wall-clock execution time**. Disabling in-process parallelism without compensating will make individual test runs slower. However:

- Helix already provides machine-level distribution for CI
- The current "fast but broken" parallel execution provides **no value** if tests are unreliable; 63% of PRs encounter intermittent test failures, so the speed is illusory

**Measured Helix compute impact (PR [#54943](https://github.com/dotnet/sdk/pull/54943)):**

*With parallelism disabled:*

| Build | Helix Work Items | Total Compute (machine-hours) |
|:------|:-----------------|:------------------------------|
| [First run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1477650) | 693 | 35.85 |
| [Second run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1478060) | 693 | 36.83 |
| [Third run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1479308) | 693 | 37.88 |
| **Average** | **693** | **36.85** |

*Baseline (parallelism enabled, 5 recent green PRs):*

| PR | Helix Work Items | Total Compute (machine-hours) |
|:---|:-----------------|:------------------------------|
| [#54859](https://github.com/dotnet/sdk/pull/54859) | 688 | 30.74 |
| [#54792](https://github.com/dotnet/sdk/pull/54792) | 691 | 31.63 |
| [#54704](https://github.com/dotnet/sdk/pull/54704) | 691 | 32.07 |
| [#54748](https://github.com/dotnet/sdk/pull/54748) | 688 | 33.77 |
| [#54896](https://github.com/dotnet/sdk/pull/54896) | 691 | 33.65 |
| **Average** | **690** | **32.37** |

Average shows a **+4.5 hour (~14%) increase** in total Helix agent compute. This is modest given the stability benefits.

### Long-term: Decide whether to reinvest in concurrent test execution

Disabling parallelism solves the immediate crisis. The longer-term question is whether there is sufficient ROI in making the test suite concurrency-safe. This would be a sustained, ongoing investment, not a one-time fix.

If we decide the ROI is there, the recommended approach is to re-enable parallelism **on a per-assembly basis**. Some test assemblies are likely better authored for concurrency than others. Rather than an all-or-nothing switch, we would evaluate each assembly individually and opt in only those that demonstrate stability under concurrent execution. This incremental approach limits risk, lets us measure the benefit of each addition, and could recover some or all of the ~14% Helix compute regression by re-enabling parallelism for the assemblies where it works reliably. For other assemblies where concurrency issues are more deeply embedded, the ROI of making them thread-safe may not justify the investment.

If we decide the ROI is not there, multi-process execution (as Roslyn does) provides an alternative path to fast local test runs without requiring concurrency-safe test code. Each process runs a test assembly in isolation, and parallelism comes from running many processes simultaneously. This would only apply to the local development experience; in CI, Helix already provides machine-level distribution across work items, so multi-process execution is not needed there.

Regardless of the parallelism decision, there are follow-up investments we can make to improve test infrastructure robustness:

- **Replace raw `GetTempPath` usage** with a test-aware alternative that provides isolation and traceability
- **Write analyzers to catch bad patterns** (e.g., shared mutable state, non-unique temp paths, unsynchronized file access) before they reach CI
- These improvements reduce future risk whether or not parallelism is eventually re-enabled

## Conclusion

Over 63% of PRs encounter intermittent test failures, and more than a third require `/ba-g` to merge. This is not a sustainable state. The underlying concurrency issues are systemic, difficult to diagnose, and intractable to fix individually. Continuing down the path of addressing these failures one by one is not enough. This problem is affecting us right now, and we need to get it under control.

Disabling in-process parallelism eliminates the root cause immediately. Once stability is restored, we can then decide whether the long-term ROI of making the test suite concurrency-safe justifies the sustained investment required, or whether alternative approaches to fast execution better serve the team.

---

<details>
<summary><h2>Appendix: PR Data</h2></summary>

**Data collected:** 2026-06-23
**Sample:** 100 most recently merged PRs into `main` (2026-06-12 to 2026-06-23)
**Method:**
- CI result: GitHub check-runs API, `dotnet-sdk-public-ci` conclusion on PR head commit
- /ba-g: GitHub issues comments API (`per_page=100`), searched for `/ba-g` substring
- Retries: Whether `/azp run` comments were posted after the initial trigger
- KBE Auto: Build Analysis check-run auto-handled known test errors (auto-dismiss or auto-retry)

| PR | CI Result | /ba-g | Retries | KBE Auto |
|:---|:----------|:------|:--------|:---------|
| [#54936](https://github.com/dotnet/sdk/pull/54936) | failure | Yes | No | No |
| [#54918](https://github.com/dotnet/sdk/pull/54918) | success | No | No | Yes |
| [#54908](https://github.com/dotnet/sdk/pull/54908) | failure | Yes | No | No |
| [#54905](https://github.com/dotnet/sdk/pull/54905) | failure | Yes | No | No |
| [#54903](https://github.com/dotnet/sdk/pull/54903) | failure | Yes | No | No |
| [#54902](https://github.com/dotnet/sdk/pull/54902) | success | No | No | No |
| [#54901](https://github.com/dotnet/sdk/pull/54901) | failure | Yes | No | No |
| [#54900](https://github.com/dotnet/sdk/pull/54900) | success | No | No | No |
| [#54899](https://github.com/dotnet/sdk/pull/54899) | success | No | No | Yes |
| [#54898](https://github.com/dotnet/sdk/pull/54898) | failure | Yes | No | No |
| [#54897](https://github.com/dotnet/sdk/pull/54897) | failure | Yes | No | No |
| [#54896](https://github.com/dotnet/sdk/pull/54896) | success | No | No | No |
| [#54895](https://github.com/dotnet/sdk/pull/54895) | failure | Yes | No | No |
| [#54893](https://github.com/dotnet/sdk/pull/54893) | success | No | No | No |
| [#54892](https://github.com/dotnet/sdk/pull/54892) | failure | No | No | No |
| [#54888](https://github.com/dotnet/sdk/pull/54888) | success | No | No | No |
| [#54885](https://github.com/dotnet/sdk/pull/54885) | success | No | No | No |
| [#54884](https://github.com/dotnet/sdk/pull/54884) | neutral | No | No | No |
| [#54882](https://github.com/dotnet/sdk/pull/54882) | failure | Yes | No | No |
| [#54881](https://github.com/dotnet/sdk/pull/54881) | success | No | No | No |
| [#54880](https://github.com/dotnet/sdk/pull/54880) | failure | Yes | No | No |
| [#54876](https://github.com/dotnet/sdk/pull/54876) | failure | Yes | No | No |
| [#54867](https://github.com/dotnet/sdk/pull/54867) | success | No | No | No |
| [#54860](https://github.com/dotnet/sdk/pull/54860) | failure | Yes | No | No |
| [#54859](https://github.com/dotnet/sdk/pull/54859) | success | No | No | No |
| [#54858](https://github.com/dotnet/sdk/pull/54858) | failure | Yes | No | No |
| [#54857](https://github.com/dotnet/sdk/pull/54857) | failure | Yes | No | No |
| [#54854](https://github.com/dotnet/sdk/pull/54854) | failure | Yes | No | No |
| [#54853](https://github.com/dotnet/sdk/pull/54853) | failure | Yes | No | No |
| [#54852](https://github.com/dotnet/sdk/pull/54852) | failure | Yes | No | No |
| [#54851](https://github.com/dotnet/sdk/pull/54851) | failure | Yes | No | No |
| [#54850](https://github.com/dotnet/sdk/pull/54850) | failure | Yes | No | No |
| [#54849](https://github.com/dotnet/sdk/pull/54849) | failure | Yes | No | No |
| [#54848](https://github.com/dotnet/sdk/pull/54848) | failure | Yes | No | No |
| [#54847](https://github.com/dotnet/sdk/pull/54847) | failure | Yes | No | No |
| [#54846](https://github.com/dotnet/sdk/pull/54846) | success | No | No | No |
| [#54845](https://github.com/dotnet/sdk/pull/54845) | failure | Yes | No | No |
| [#54842](https://github.com/dotnet/sdk/pull/54842) | failure | Yes | No | No |
| [#54834](https://github.com/dotnet/sdk/pull/54834) | success | No | No | No |
| [#54831](https://github.com/dotnet/sdk/pull/54831) | success | No | No | Yes |
| [#54825](https://github.com/dotnet/sdk/pull/54825) | success | No | No | No |
| [#54824](https://github.com/dotnet/sdk/pull/54824) | success | No | No | No |
| [#54820](https://github.com/dotnet/sdk/pull/54820) | success | No | No | No |
| [#54810](https://github.com/dotnet/sdk/pull/54810) | success | No | No | No |
| [#54807](https://github.com/dotnet/sdk/pull/54807) | failure | Yes | No | No |
| [#54804](https://github.com/dotnet/sdk/pull/54804) | failure | Yes | No | No |
| [#54800](https://github.com/dotnet/sdk/pull/54800) | success | No | No | No |
| [#54799](https://github.com/dotnet/sdk/pull/54799) | success | No | No | Yes |
| [#54797](https://github.com/dotnet/sdk/pull/54797) | success | No | No | Yes |
| [#54794](https://github.com/dotnet/sdk/pull/54794) | success | No | No | No |
| [#54793](https://github.com/dotnet/sdk/pull/54793) | success | No | No | No |
| [#54792](https://github.com/dotnet/sdk/pull/54792) | success | No | No | No |
| [#54791](https://github.com/dotnet/sdk/pull/54791) | success | No | No | No |
| [#54790](https://github.com/dotnet/sdk/pull/54790) | success | No | No | No |
| [#54789](https://github.com/dotnet/sdk/pull/54789) | success | No | No | No |
| [#54787](https://github.com/dotnet/sdk/pull/54787) | success | No | No | No |
| [#54785](https://github.com/dotnet/sdk/pull/54785) | failure | Yes | Yes | No |
| [#54782](https://github.com/dotnet/sdk/pull/54782) | success | No | No | Yes |
| [#54780](https://github.com/dotnet/sdk/pull/54780) | failure | Yes | No | No |
| [#54772](https://github.com/dotnet/sdk/pull/54772) | success | No | No | No |
| [#54771](https://github.com/dotnet/sdk/pull/54771) | success | No | No | Yes |
| [#54770](https://github.com/dotnet/sdk/pull/54770) | success | No | No | Yes |
| [#54769](https://github.com/dotnet/sdk/pull/54769) | success | No | No | No |
| [#54768](https://github.com/dotnet/sdk/pull/54768) | success | No | No | No |
| [#54767](https://github.com/dotnet/sdk/pull/54767) | success | No | No | No |
| [#54766](https://github.com/dotnet/sdk/pull/54766) | success | No | No | No |
| [#54765](https://github.com/dotnet/sdk/pull/54765) | failure | Yes | Yes | No |
| [#54761](https://github.com/dotnet/sdk/pull/54761) | failure | Yes | Yes | No |
| [#54759](https://github.com/dotnet/sdk/pull/54759) | failure | Yes | No | No |
| [#54758](https://github.com/dotnet/sdk/pull/54758) | success | No | Yes | No |
| [#54755](https://github.com/dotnet/sdk/pull/54755) | neutral | No | No | No |
| [#54752](https://github.com/dotnet/sdk/pull/54752) | failure | Yes | No | No |
| [#54749](https://github.com/dotnet/sdk/pull/54749) | success | No | Yes | No |
| [#54748](https://github.com/dotnet/sdk/pull/54748) | success | No | No | No |
| [#54747](https://github.com/dotnet/sdk/pull/54747) | success | No | No | Yes |
| [#54745](https://github.com/dotnet/sdk/pull/54745) | success | No | No | Yes |
| [#54744](https://github.com/dotnet/sdk/pull/54744) | success | No | No | No |
| [#54743](https://github.com/dotnet/sdk/pull/54743) | success | No | No | No |
| [#54741](https://github.com/dotnet/sdk/pull/54741) | success | No | No | No |
| [#54732](https://github.com/dotnet/sdk/pull/54732) | success | No | Yes | No |
| [#54731](https://github.com/dotnet/sdk/pull/54731) | success | No | No | Yes |
| [#54729](https://github.com/dotnet/sdk/pull/54729) | failure | Yes | Yes | No |
| [#54727](https://github.com/dotnet/sdk/pull/54727) | neutral | No | No | No |
| [#54726](https://github.com/dotnet/sdk/pull/54726) | failure | Yes | Yes | No |
| [#54725](https://github.com/dotnet/sdk/pull/54725) | failure | Yes | Yes | No |
| [#54724](https://github.com/dotnet/sdk/pull/54724) | success | No | Yes | No |
| [#54716](https://github.com/dotnet/sdk/pull/54716) | success | No | No | Yes |
| [#54713](https://github.com/dotnet/sdk/pull/54713) | failure | Yes | No | No |
| [#54709](https://github.com/dotnet/sdk/pull/54709) | success | No | No | Yes |
| [#54708](https://github.com/dotnet/sdk/pull/54708) | success | No | No | Yes |
| [#54707](https://github.com/dotnet/sdk/pull/54707) | success | No | No | No |
| [#54705](https://github.com/dotnet/sdk/pull/54705) | success | No | Yes | Yes |
| [#54704](https://github.com/dotnet/sdk/pull/54704) | success | No | No | No |
| [#54703](https://github.com/dotnet/sdk/pull/54703) | success | No | No | Yes |
| [#54702](https://github.com/dotnet/sdk/pull/54702) | success | No | No | No |
| [#54701](https://github.com/dotnet/sdk/pull/54701) | success | No | Yes | No |
| [#54700](https://github.com/dotnet/sdk/pull/54700) | success | No | No | No |
| [#54699](https://github.com/dotnet/sdk/pull/54699) | success | No | No | Yes |
| [#54698](https://github.com/dotnet/sdk/pull/54698) | success | No | No | Yes |
| [#54693](https://github.com/dotnet/sdk/pull/54693) | success | No | No | No |

</details>

### PR DESCRIPTION
**Pull request by MichaelSimons**
## Purpose

Disable in-process parallel test execution to address widespread concurrency-related intermittent test failures.

## Changes

- `test/xunit.runner.json`: Set `parallelizeAssembly: false` and `parallelizeTestCollections: false`
- `test/TemplateEngine/.../xunit.runner.json`: Same
- `test/Directory.Build.props`: Set `MSTestParallelizeScope=None`

## Context

See #54967 for the full analysis and data supporting this change.

Closes #54967

### CODE REVIEW

File: test/Directory.Build.props [hunk 1]
```diff
@@ -28,9 +28,9 @@
   <!-- MSTest.Sdk defaults. Applied to projects that opt in via <Project Sdk="MSTest.Sdk">,
        which sets UsingMSTestSdk=true in its Sdk.props before this file is evaluated. -->
   <PropertyGroup Condition="'$(UsingMSTestSdk)' == 'true'">
-    <!-- Run tests in parallel at the method level by default. MSTest.TestAdapter
-         emits the corresponding [assembly: Parallelize(Scope = MethodLevel)] attribute. -->
-    <MSTestParallelizeScope>MethodLevel</MSTestParallelizeScope>
+    <!-- Parallelization disabled for investigation — scope=None prevents MSTest from
+         injecting [Parallelize], avoiding conflicts with [DoNotParallelize] attributes. -->
```
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply test/Directory.Build.props]: fixed

**Review by joeloff — APPROVED** [review]

**Review by Evangelink — APPROVED** [review]
